src/ring_pedersen.rs

```diff
@@ -0,0 +1,220 @@
+use crate::{
+    errors::Result,
+    utils::modpow,
+    zkp::{
+        piprm::{PiPrmInput, PiPrmProof, PiPrmSecret},
+        Proof,
+    },
+};
+use crate::{paillier::DecryptionKey, utils::random_plusminus};
+use libpaillier::unknown_order::BigNumber;
+use rand::{CryptoRng, RngCore};
+use serde::{Deserialize, Serialize};
+
+/// A commitment scheme based on a ring-variant of the Pedersen commitment scheme.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub(crate) struct RingPedersen {
+    /// The RSA modulus, corresponding to `N` in the paper.
+    modulus: BigNumber,
+    /// Ring-Petersen parameter `s`.
+    s: BigNumber,
+    /// Ring-Petersen parameter `t`.
+    t: BigNumber,
+}
+
+/// A commitment produced by [`RingPedersen::commit`].
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+pub(crate) struct Commitment(BigNumber);
+
+impl Commitment {
+    pub(crate) fn to_bytes(&self) -> Vec<u8> {
+        self.0.to_bytes()
+    }
+}
+
+/// The randomness generated as part of [`RingPedersen::commit`].
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub(crate) struct CommitmentRandomness(BigNumber);
+
+impl CommitmentRandomness {
+    pub(crate) fn to_bytes(&self) -> Vec<u8> {
+        self.0.to_bytes()
+    }
+
+    /// Masks value with `mask` and `challenge`.
+    /// The output [`MaskedRandomness`] value is computed as `mask - challenge * value`.
+    pub(crate) fn mask_neg(
+        &self,
+        mask: &CommitmentRandomness,
+        challenge: &BigNumber,
+    ) -> MaskedRandomness {
+        MaskedRandomness(&mask.0 - challenge * &self.0)
+    }
+
+    /// Masks value with `mask` and `challenge`.
+    /// The output [`MaskedRandomness`] value is computed as `mask + challenge * value`.
+    pub(crate) fn mask(
+        &self,
+        mask: &CommitmentRandomness,
+        challenge: &BigNumber,
+    ) -> MaskedRandomness {
+        MaskedRandomness(&mask.0 + challenge * &self.0)
+    }
+}
+
+/// The randomness generated as part of [`RingPedersen::commit`] and masked
+/// by [`CommitmentRandomness::mask`] or [`CommitmentRandomness::mask_neg`].
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub(crate) struct MaskedRandomness(BigNumber);
+
+impl From<CommitmentRandomness> for MaskedRandomness {
+    fn from(value: CommitmentRandomness) -> Self {
+        Self(value.0)
+    }
+}
+
+impl From<MaskedRandomness> for CommitmentRandomness {
+    fn from(value: MaskedRandomness) -> Self {
+        Self(value.0)
+    }
+}
```

Comment:
suggestion: I think I saw in the usage that you downgrade masked -> commitment randomness and then immediately mask it again. Can we just make a `mask` or `remask` method on `MaskedRandomness` instead?